cln_plugin/server.go

```diff
@@ -0,0 +1,432 @@
+// the server is done. This function effectively waits until a subscriber is
+// active and has sent a message.
+func (s *server) Receive() (string, interface{}) {
+	select {
+	case <-s.done:
+		return "", nil
+	case msg := <-s.recvQueue:
+		return msg.id, msg.result
+	}
+}
+
+// Listens to sendQueue for htlc_accepted requests from cln. The message will be
+// held until a subscriber is active, or the subscriber timeout expires. The
+// messages are sent to the grpc client in fifo order.
+func (s *server) listenHtlcRequests() {
+	for {
+		select {
+		case <-s.done:
+			log.Printf("listenHtlcRequests received done. Stop listening.")
+			return
+		case msg := <-s.sendQueue:
+			s.handleHtlcAccepted(msg)
+		}
+	}
+}
+
+// Attempts to send a htlc_accepted message to the grpc client. The message will
+// be held until a subscriber is active, or the subscriber timeout expires.
+func (s *server) handleHtlcAccepted(msg *htlcAcceptedMsg) {
+	for {
+		s.mtx.Lock()
+		sb := s.subscription
+		ns := s.newSubscriber
+		s.mtx.Unlock()
+
+		// If there is no active subscription, wait until there is a new
+		// subscriber, or the message times out.
+		if sb == nil {
+			select {
+			case <-s.done:
+				log.Printf("handleHtlcAccepted received server done. Stop processing.")
+				return
+			case <-ns:
+				log.Printf("got a new subscriber. continue handleHtlcAccepted.")
+				continue
+			case <-time.After(time.Until(msg.timeout)):
+				log.Printf(
+					"WARNING: htlc with id '%s' timed out after '%v' waiting "+
+						"for grpc subscriber: %+v",
+					msg.id,
+					s.subscriberTimeout,
+					msg.htlc,
+				)
+				s.recvQueue <- &htlcResultMsg{
+					id:     msg.id,
+					result: s.defaultResult(),
+				}
+
+				return
+			}
+		}
+
+		// There is a subscriber. Attempt to send the htlc_accepted message.
+		err := sb.stream.Send(&proto.HtlcAccepted{
+			Correlationid: msg.id,
+			Onion: &proto.Onion{
+				Payload:           msg.htlc.Onion.Payload,
+				ShortChannelId:    msg.htlc.Onion.ShortChannelId,
+				ForwardMsat:       msg.htlc.Onion.ForwardMsat,
+				OutgoingCltvValue: msg.htlc.Onion.OutgoingCltvValue,
+				SharedSecret:      msg.htlc.Onion.SharedSecret,
+				NextOnion:         msg.htlc.Onion.NextOnion,
+			},
+			Htlc: &proto.Htlc{
+				ShortChannelId:     msg.htlc.Htlc.ShortChannelId,
+				Id:                 msg.htlc.Htlc.Id,
+				AmountMsat:         msg.htlc.Htlc.AmountMsat,
+				CltvExpiry:         msg.htlc.Htlc.CltvExpiry,
+				CltvExpiryRelative: msg.htlc.Htlc.CltvExpiryRelative,
+				PaymentHash:        msg.htlc.Htlc.PaymentHash,
+			},
+			ForwardTo: msg.htlc.ForwardTo,
+		})
+
+		// If there is no error, we're done.
+		if err == nil {
+			return
+		}
+
+		// If we end up here, there was an error sending the message to the
+		// grpc client.
+		log.Printf("Error sending htlc_accepted message to subscriber. "+
+			"Removing subscription: %v", err)
+		s.removeSubscriptionIfUnchanged(sb, err)
```

Comment:
Do we need this code? After all the code already removes the subscription when the stream is closed.

Reply:
I don't really know. That depends on whether the HtlcStream context is done whenever there's a send error I guess. I can look into that.

Reply:
In theory there could be a send error that won't close the stream and leave the context alive but do we  need to remove the subscription in such case? I would say no. Coupling the subscriber to the stream seems the right and simple way to me.

Reply:
I was thinking if the connection is gone you want to remove that subscriber asap. Because if there is a dead subscriber hanging in the plugin side, lspd cannot connect to the plugin until that subscriber is removed. Having a maximum of 1 subscriber I think is a good thing, to avoid weird cases with duplicate broadcasting of htlc_accepted, or missing a broadcast somewhere. So whenever there is a chance to remove the subscriber earlier I think we should take it. LSPD will reconnect promptly if the connection drops.

Reply:
I agree 1 subscriber is good.
if we bind the subscription to the stream (which requires live tcp connection) how can we have a "dead subscriber"?
In practice I think  when you get error from the Send it means the stream is closed already which probably means  the subscription already being removed.

Reply:
By default the grpc api would allow multiple simultaneous subscriptions. In order to allow only one subscription, the plugin has to keep some kind of state. Could be a boolean, the stream, or an object (like the subscription object I coded now).

From what I remember, TCP connections are weird. The client could get an error if it's no longer connected to the server, while the server is unaware for a while (probably because a keepalive ping will fail after some time the connection will finally drop). lspd and the plugin are both server and client in this case. It's this period of inconsistency in knowledge about the connection state that made me code up dropping the connection whenever something fails, so at least you're sure the plugin will not reject a new connection is the server is aware the connection has dropped already.

Reply:
I see the intention but since you get the error in the plugin it is a clear indication that the connection has dropped in the plugin side which is what you are interested in.
If the other side still hanging on the closed connection from some reason then the explicit close won't make a difference.
This is why I think the code could be simplified a bit.

Reply:
I'm going to think about a way to test that behavior of removing the subscriber / subscribing to context cancellation. That will take some time I think. lspd itself currently doesn't exit in my integration tests. I think that is because the `Recv` call never returns if the server is gone. You'd hope the context cancels then, but it might not be so obvious.

Reply:
The Receive should return when the connection is closed which may take some time for the client  to detect (as you mentioned previously). A good way to detect such closing early is to configure frequent keep alive (https://pkg.go.dev/google.golang.org/grpc/keepalive).
If the keep alive is not answered in a certain amount of time then the connection would be closed on that side, worth trying it and I think we can be aggressive with the keep alive as we only have one subscriber.